### PR TITLE
[CARBONDATA-1626]add data size and index size in table status file

### PR DESCRIPTION
**Pull request by akashrn5**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[CARBONDATA-<Jira issue #>] Description of pull request`
   
 - [ ] Make sure to add PR description including
        
        - the root cause/problem statement
        - What is the implemented solution

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
 
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
         
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
                 
---

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/675/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1312/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/678/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1315/

**Comment by akashrn5** [comment]
@jackylk please review

**Comment by akashrn5** [comment]
@gvramana please review

**Comment by akashrn5** [comment]
@ravipesala please review

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 1]
```diff
@@ -1376,6 +1376,32 @@
...
+
+  /**
+   * ENABLE_CALCULATE_DATA_INDEX_SIZE
+   */
+  @CarbonProperty public static final String ENABLE_CALCULATE_SIZE = "carbon.enable.calculate.size";
+
+  /**
+   * DEFAULT_ENABLE_CALCULATE_DATA_INDEX_SIZE
+   */
+  @CarbonProperty public static final String DEFAULT_ENABLE_CALCULATE_SIZE = "true";
```
Comment (mohammadshahidkhan) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: for constant variable/keys CarbonProperty anotation not required

Comment (akashrn5) [reply core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: ok

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1115/

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java [hunk 2]
```diff
@@ -509,7 +512,7 @@ public static boolean createNewLockFile(String filePath, FileType fileType) thro
   }
 
   public enum FileType {
-    LOCAL, HDFS, ALLUXIO, VIEWFS
+    LOCAL, HDFS, ALLUXIO, VIEWFS, S3
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java]: S3 is not supported yet.

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java [hunk 3]
```diff
@@ -606,4 +609,53 @@ public static FileSystem getFileSystem(Path path) throws IOException {
     return path.getFileSystem(configuration);
   }
 
+  // Get the total size of carbon data and the total size of carbon index
+  public static HashMap<String, Long> getDataSizeAndIndexSize(CarbonTablePath carbonTablePath,
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java]: This method cannot be in FileFactory, as it in not generic filesystem operation.

File: core/src/main/java/org/apache/carbondata/core/memory/HeapMemoryAllocator.java [hunk 4]
```diff
@@ -17,11 +17,11 @@
 
 package org.apache.carbondata.core.memory;
 
-import javax.annotation.concurrent.GuardedBy;
 import java.lang.ref.WeakReference;
 import java.util.HashMap;
 import java.util.LinkedList;
 import java.util.Map;
+import javax.annotation.concurrent.GuardedBy;
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/memory/HeapMemoryAllocator.java]: why is this added?

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1714/

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java (hunk 3 above)
Comment (akashrn5) [reply core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java]: handled comments

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1191/

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 5]
```diff
@@ -2119,5 +2127,146 @@ public static String getNewTablePath(Path carbonTablePath,
...
+    loadMetadataDetails
+        .setDataSize(dataIndexSize.get(CarbonCommonConstants.CARBON_TOTAL_DATA_SIZE).toString());
+    loadMetadataDetails
+        .setIndexSize(dataIndexSize.get(CarbonCommonConstants.CARBON_TOTAL_INDEX_SIZE).toString());
+  }
+
+  /**
+   * This method will calculate the data size and index size for carbon table
+   */
+  public static HashMap<String, Long> calculateSize(CarbonTable carbonTable)
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: Update the method signature Map<String, Long>

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 6]
```diff
@@ -2119,5 +2127,146 @@ public static String getNewTablePath(Path carbonTablePath,
...
   }
+
+  /*
+   * This method will add data size and index size into tablestatus for each segment
+   */
+  public static void addDataIndexSizeIntoMetaEntry(LoadMetadataDetails loadMetadataDetails,
+      String segmentId, CarbonTable carbonTable) throws IOException {
+    CarbonTablePath carbonTablePath =
+        CarbonStorePath.getCarbonTablePath((carbonTable.getAbsoluteTableIdentifier()));
+    HashMap<String, Long> dataIndexSize =
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: Change it to Map<String,Long>

File: integration/spark2/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 7]
```diff
@@ -292,6 +290,35 @@ object CarbonDataRDDFactory {
     var executorMessage: String = ""
     val isSortTable = carbonTable.getNumberOfSortColumns > 0
     val sortScope = CarbonDataProcessorUtil.getSortScope(carbonLoadModel.getSortScope)
+
+    def updateStatus(status: Array[(String, (LoadMetadataDetails, ExecutionErrors))],
```
Comment (kumarvishal09) [inline integration/spark2/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: do not update table status file separately in separate method for size, add  size while adding loadmetadata details to table status

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1196/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1221/

File: integration/spark2/src/test/scala/org/apache/spark/sql/GetDataSizeAndIndexSizeTest.scala [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.test.util.QueryTest
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.scalatest.BeforeAndAfterAll
+
+class GetDataSizeAndIndexSizeTest extends QueryTest with BeforeAndAfterAll {
```
Comment (ravipesala) [inline integration/spark2/src/test/scala/org/apache/spark/sql/GetDataSizeAndIndexSizeTest.scala]: Please add testcase of update scenerio

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1227/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1232/

**Comment by ravipesala** [comment]
LGTM
